Recompute event target when removed from changing selection

Ahmad-S792 opened the pull request:
#### 18561227325c21d4b156974983f0f44da1979a45
<pre>
Recompute event target when removed from changing selection

<a href="https://bugs.webkit.org/show_bug.cgi?id=284205">https://bugs.webkit.org/show_bug.cgi?id=284205</a>
<a href="https://rdar.apple.com/141080146">rdar://141080146</a>

Reviewed by Wenson Hsieh.

Merge: <a href="https://chromium.googlesource.com/chromium/blink/+/f2696796968c131cf121536fd8c5f87d801d90a7">https://chromium.googlesource.com/chromium/blink/+/f2696796968c131cf121536fd8c5f87d801d90a7</a>

Changing selection by mouse click uses event target node to calculate position. This processing is
performed after event dispatch. When event handler removes event target node, selection isn&apos;t changed.

This patch changes this behavior to pass recomputing event target when it is removed for changing selection.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::shouldRefetchEventTarget):
(WebCore::EventHandler::handleMousePressEvent):
* LayoutTests/editing/selection/caret-after-removing-node.html: Add Test Case
* LayoutTests/editing/selection/caret-after-removing-node-expected.txt: Add Test Case Expectation
* LayoutTests/platform/ios/TestExpectations: Skipped on iOS due to lack of MouseEvent

Canonical link: <a href="https://commits.webkit.org/287619@main">https://commits.webkit.org/287619@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0736b5d5ac6b60ddf697258068c48f8545ba7f83

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80313 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/59319 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33737 "Failed to checkout and rebase branch from PR 37576") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84834 "Failed to checkout and rebase branch from PR 37576") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/31295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82424 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68381 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7619 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/84834 "Failed to checkout and rebase branch from PR 37576") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/31295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83382 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52850 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/55/builds/33737 "Failed to checkout and rebase branch from PR 37576") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/84834 "Failed to checkout and rebase branch from PR 37576") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/50173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/55/builds/33737 "Failed to checkout and rebase branch from PR 37576") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29754 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71310 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/55/builds/33737 "Failed to checkout and rebase branch from PR 37576") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86268 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7538 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/71065 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7713 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/55/builds/33737 "Failed to checkout and rebase branch from PR 37576") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/70305 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17500 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/14301 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/55/builds/33737 "Failed to checkout and rebase branch from PR 37576") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7500 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7339 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/10859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9145 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -380,6 +380,20 @@ bool EventHandler::eventLoopHandleMouseDragged(const MouseEventWithHitTestResult
+{
+    RefPtr targetNode = mouseEvent.targetNode();
+    RefPtr shadowRoot = dynamicDowncast<ShadowRoot>(*targetNode);
+    ASSERT(targetNode);
```
Nit - this assertion should either be removed or moved up 1 line, since `targetNode` is already dereferenced by the time we reach this assert.

Ahmad-S792 replied:
Can do once EWS finishes. :-)

Ahmad-S792 replied:
@whsieh - done. :-)

whsieh commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -380,6 +380,20 @@ bool EventHandler::eventLoopHandleMouseDragged(const MouseEventWithHitTestResult
+    ASSERT(targetNode);
+    RefPtr shadowRoot = dynamicDowncast<ShadowRoot>(*targetNode);
+    if (!targetNode->parentNode())
+        return true;
```
I'd probably invert the order here, to prevent an unnecessary `dynamicDowncast` in the case where `targetNode` has no parent.

```suggestion
    if (!targetNode->parentNode())
        return true;
    RefPtr shadowRoot = dynamicDowncast<ShadowRoot>(*targetNode);
```

Ahmad-S792 replied:
Sure - I will update it. Thanks!

Ahmad-S792 replied:
Done.

Ahmad-S792 committed:
Recompute event target when removed from changing selection

https://bugs.webkit.org/show_bug.cgi?id=284205
rdar://141080146

Reviewed by Wenson Hsieh.

Merge: https://chromium.googlesource.com/chromium/blink/+/f2696796968c131cf121536fd8c5f87d801d90a7

Changing selection by mouse click uses event target node to calculate position. This processing is
performed after event dispatch. When event handler removes event target node, selection isn't changed.

This patch changes this behavior to pass recomputing event target when it is removed for changing selection.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::shouldRefetchEventTarget):
(WebCore::EventHandler::handleMousePressEvent):
* LayoutTests/editing/selection/caret-after-removing-node.html: Add Test Case
* LayoutTests/editing/selection/caret-after-removing-node-expected.txt: Add Test Case Expectation
* LayoutTests/platform/ios/TestExpectations: Skipped on iOS due to lack of MouseEvent

Canonical link: https://commits.webkit.org/287619@main
